cardano-client-demo/Main.hs

```diff
@@ -0,0 +1,98 @@
+import System.FilePath ((</>))
+
+-- | Connects to a local cardano node, requests the blocks and prints out the
+-- number of transactions. To run this, you must first start a local node e.g.:
+--
+--     $ cabal run cardano-node:cardano-node -- run \
+--        --config        configuration/cardano/mainnet-config.json \
+--        --topology      configuration/cardano/mainnet-topology.json \
+--        --database-path db \
+--        --socket-path   db/node.sock \
+--        --host-addr     127.0.0.1 \
+--        --port          3001 \
+--
+-- Then run this with the path to the directory containing node.sock:
+--
+--     $ cabal run cardano-client-demo -- db
+--
+main :: IO ()
+main = do
+  -- Get cocket path from CLI argument.
+  socketDir:_ <- getArgs
+  let socketPath = socketDir </> "node.sock"
+
+  -- Connect to the node.
+  putStrLn $ "Connecting to socket: " <> socketPath
+  IPC.connectToLocalNode
+    (connectInfo socketPath)
+    protocols
+  where
+  connectInfo :: FilePath -> IPC.LocalNodeConnectInfo IPC.CardanoMode
+  connectInfo socketPath =
+      IPC.LocalNodeConnectInfo {
+        IPC.localConsensusModeParams = IPC.CardanoModeParams (EpochSlots 21600),
+        IPC.localNodeNetworkId       = Mainnet,
+        IPC.localNodeSocketPath      = socketPath
+      }
+
+  protocols :: IPC.LocalNodeClientProtocolsInMode IPC.CardanoMode
+  protocols =
+      IPC.LocalNodeClientProtocols {
+        IPC.localChainSyncClient    = Just chainSyncClient,
+        IPC.localTxSubmissionClient = Nothing,
+        IPC.localStateQueryClient   = Nothing
+      }
+
+-- | Defines the client side of the chain sync protocol.
+chainSyncClient :: ChainSyncClient
+                     (IPC.BlockInMode IPC.CardanoMode)
+                     ChainPoint
+                     ChainTip
+                     IO ()
+chainSyncClient = ChainSyncClient clientStIdle
+  where
+    clientStIdle :: IO (ClientStIdle (IPC.BlockInMode IPC.CardanoMode)
+                                 ChainPoint ChainTip IO ())
+    clientStIdle = do
+      putStrLn "Chain Sync: requesting next"
+      return $ SendMsgRequestNext
+        -- There's more to get immediately
+        clientStNext
+
+        -- The node is asking us to wait. This is because we reached the
+        -- tip. We can certainly carry on here, but for this demo we are
+        -- going to stop when we hit the current chain tip.
+        clientDone
+
+    clientStNext :: ClientStNext (IPC.BlockInMode IPC.CardanoMode)
+                                 ChainPoint ChainTip IO ()
+    clientStNext =
+      ClientStNext {
+          recvMsgRollForward = \(IPC.BlockInMode (Block _header transactions) _) _tip ->
+            ChainSyncClient $ do
+              putStrLn $ "Block # of Transactions:  " ++ show (length transactions)
+              clientStIdle
+
+        , recvMsgRollBackward = \_ _ -> ChainSyncClient clientStIdle
+        }
+
+    -- We're still in the "Next" state here, but we've decided to stop
```

Comment:
Does this mean client still need to wait for the next block?

Reply:
Indeed. The types keep track of the number of requests in flight. We can only terminate the protocol when 0 requests are in flight. Hence the `Z` meaning 0 in `SendMsgDone :: a -> ClientPipelinedStIdle Z header point tip m a`